vars/withDockerEnv.groovy

```diff
@@ -0,0 +1,73 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+  Configure the Docker context to run the body closure, login to hub.docker.com with an
+  authentication credentials from a Vault secret. The vault secret contains `user` and `password`
+  fields with the authentication details.
+
+  withDockerEnv() {
+    // block
+  }
+  withDockerEnv(secret: 'secret/team/ci/secret-name') {
+    // block
+  }
+  withDockerEnv(secret: 'secret/team/ci/secret-name', registry: "docker.io") {
+    // block
+  }
+
+*/
+def call(Map args = [:], Closure body){
+  def secret = args.containsKey('secret') ? args.secret : error("withDockerEnv: No valid secret to looking for.")
```

Comment:
I'd like to deprecate the `dockerLogin` calling this step inside it. But the unit test are failing locally if I replace the implementation of that step:

```groovy
def call(Map args = [:]){
  withDockerEnv(args) {
    // NOOP
  }
}
```

```shell
$ ./mvnw clean test -Dtest=DockerLoginStepTests -Pdebug
...
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running DockerLoginStepTests
[ERROR] Tests run: 5, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 1.211 s <<< FAILURE! - in DockerLoginStepTests
[ERROR] DockerLoginStepTests.test  Time elapsed: 0.937 s  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: dockerLogin.withDockerEnv() is applicable for argument types: (java.util.LinkedHashMap, dockerLogin$_call_closure1) values: [[secret:secret/team/ci/secret-name], dockerLogin$_call_closure1@64c63c79]
        at DockerLoginStepTests.test(DockerLoginStepTests.groovy:33)

[ERROR] DockerLoginStepTests.testRegistry  Time elapsed: 0.062 s  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: dockerLogin.withDockerEnv() is applicable for argument types: (java.util.LinkedHashMap, dockerLogin$_call_closure1) values: [[secret:secret/team/ci/secret-name, registry:other.docker.io], ...]
        at DockerLoginStepTests.testRegistry(DockerLoginStepTests.groovy:41)

[ERROR] DockerLoginStepTests.testRegistryInWindows  Time elapsed: 0.064 s  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: dockerLogin.withDockerEnv() is applicable for argument types: (java.util.LinkedHashMap, dockerLogin$_call_closure1) values: [[secret:secret/team/ci/secret-name, registry:other.docker.io], ...]
        at DockerLoginStepTests.testRegistryInWindows(DockerLoginStepTests.groovy:59)

[ERROR] DockerLoginStepTests.testWindows  Time elapsed: 0.048 s  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: dockerLogin.withDockerEnv() is applicable for argument types: (java.util.LinkedHashMap, dockerLogin$_call_closure1) values: [[secret:secret/team/ci/secret-name], dockerLogin$_call_closure1@40db2a24]
        at DockerLoginStepTests.testWindows(DockerLoginStepTests.groovy:50)

[ERROR] DockerLoginStepTests.test_with_role_secret  Time elapsed: 0.042 s  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: dockerLogin.withDockerEnv() is applicable for argument types: (java.util.LinkedHashMap, dockerLogin$_call_closure1) values: [[secret:secret/team/ci/secret-name, role_id:role-id, secret_id:secret-id], ...]
        at DockerLoginStepTests.test_with_role_secret(DockerLoginStepTests.groovy:67)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   DockerLoginStepTests.test:33 » MissingMethod No signature of method: dockerLog...
[ERROR]   DockerLoginStepTests.testRegistry:41 » MissingMethod No signature of method: d...
[ERROR]   DockerLoginStepTests.testRegistryInWindows:59 » MissingMethod No signature of ...
[ERROR]   DockerLoginStepTests.testWindows:50 » MissingMethod No signature of method: do...
[ERROR]   DockerLoginStepTests.test_with_role_secret:67 » MissingMethod No signature of ...
[INFO] 
[ERROR] Tests run: 5, Failures: 0, Errors: 5, Skipped: 0
```

Reply:
Declare the `withDockerEnv()` in `src/test/groovy/ApmBasePipelineTest.groovy` under the `registerSharedLibraryMethods` method.

Reply:
@v1v I added it here: https://github.com/elastic/apm-pipeline-library/pull/1306/commits/5e4ad726e41e305b43ea08d976c047c277e5bdef

Do you think it's a good idea?